sandbox/apparmor/notify/ioctl.go

```diff
@@ -0,0 +1,142 @@
+package notify
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"log"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/snapcore/snapd/arch"
+	"github.com/snapcore/snapd/osutil"
+)
+
+var ErrIoctlReturnInvalid = errors.New("IOCTL request returned invalid bufsize")
+
+var doSyscall = func(trap, a1, a2, a3 uintptr) (r1, r2 uintptr, err unix.Errno) {
+	return unix.Syscall(trap, a1, a2, a3)
+}
+
+type hexBuf []byte
+
+func (hb hexBuf) String() string {
+	var buf bytes.Buffer
+	for i, b := range hb {
+		if i%16 == 0 && i > 0 {
+			fmt.Fprintf(&buf, "\n")
+		}
+		fmt.Fprintf(&buf, "%#02x", b)
+		if i != len(hb)-1 {
+			fmt.Fprintf(&buf, ", ")
+		}
+	}
+	return buf.String()
+}
+
+type IoctlRequestBuffer struct {
+	bytes []byte
+}
+
+// NewIoctlRequestBuffer returns a new buffer for communication with the kernel.
+// The buffer contains encoded information about its size and protocol version.
+func NewIoctlRequestBuffer() *IoctlRequestBuffer {
+	bufSize := 0xFFFF
+	buf := bytes.NewBuffer(make([]byte, 0, bufSize))
+	header := MsgHeader{Version: 2, Length: uint16(bufSize)}
+	order := arch.Endian()
+	binary.Write(buf, order, &header)
+	buf.Write(make([]byte, bufSize-buf.Len()))
+	return &IoctlRequestBuffer{
+		bytes: buf.Bytes(),
+	}
+}
+
+func (b *IoctlRequestBuffer) Bytes() []byte {
+	return b.bytes
+}
+
+func (b *IoctlRequestBuffer) Len() int {
+	return len(b.bytes)
+}
+
+func (b *IoctlRequestBuffer) Pointer() uintptr {
+	return uintptr(unsafe.Pointer(&b.bytes[0]))
+}
+
+var dumpIoctl bool = osutil.GetenvBool("SNAPD_DEBUG_DUMP_IOCTL")
+
+// NotifyIoctl performs a ioctl(2) on the given file descriptor.
+// Sets the length of buf.Bytes() to be equal to the return value of the
+// syscall, indicating how many bytes were written to the buffer.
+// If the ioctl syscall returns an error, the buffer contents are those filled
+// by the syscall, but the size of the buffer is unchanged, and NotifySyscall
+// returns the size and error returned by the syscall.
+func NotifyIoctl(fd uintptr, req IoctlRequest, buf *IoctlRequestBuffer) (int, error) {
```

Comment:
nitpick: with the new package name and the golang conventions this could be just called Ioctl